### PR TITLE
Add Typing Environments

### LINKED ISSUE #76: Proper Type Environments, Binders, and Assignments
Adding this as a single issue, because it addresses the following bug:

Because the statement in the inner loop should have been an assignment, but instead it is a shadowing declaration, therefore renamer uncouples them.  Furthermore, as we do not have proper binding rules for the for loop, type checker happily accepts this program, however the loop never gets evaluated, therefore the variable is never bound!

```
*** Failed! (after 646 tests and 23 shrinks):
        ERROR
          evaluation of the renamed program did not match the evaluation of the original program

        INITIAL STATE
          Env {unEnv = fromList [(26,Some 0)]}

        ORIGINAL PROGRAM
          for z_24 = 0 to -1 do
          let x_26 = x_26;
          end;
          let y_29 = if (0 == x_26) then 0 else 0;

        ORIGINAL PROGRAM FINAL STATE
          "[Some 0]"

        RENAMED PROGRAM
          for z_27 = 0 to -1 do
          let x_28 = x_26;
          end;
          let y_29 = if (0 == x_28) then 0 else 0;

        RENAMED PROGRAM FINAL STATE
          "VariableNotInScopeEvalError"
```

### PR DESCRIPTION
**Pull request by jakzale**
Adding typing environemnts

- [X] Add type environments to type checker,
- [X] Update existing tests,
- [X] Update `progToString` to generate `ext` keywords,
- [X] Add scoping rules for `for` loops.
- [x] add a `Program` section specifying all `ext` variables.
- [x] Update API `freeVars` to `extVars`,
  these are not the _free_ variables that you are looking for.
- [x] Test for loops scopes
  At the moment there is no test case for for loop shadowing an `ext` variable.

### CODE REVIEW

**Comment by jakzale** [comment]
This should fix some problems with #76, but it doesnt have assignments yet.

File: field/TinyLang/Field/Evaluator.hs [hunk 1]
```diff
@@ -323,24 +321,3 @@ normExpr(EAppBinOp op e1 e2) = do
...
--- | Instantiate some of the variables of a statement with values.
-instStatement :: Monad m => Statement f -> EvalT f m (Statement f)
-instStatement (ELet uniVar def)             = ELet uniVar <$> instExpr def
-instStatement (EAssert expr)                = EAssert <$> instExpr expr
-
-instStatements :: Monad m => Statements f -> EvalT f m (Statements f)
-instStatements = traverse instStatement
-
--- | Instantiate some of the variables of an expression with values.
-instExpr :: Monad m => Expr f a -> EvalT f m (Expr f a)
```
Comment (effectfully) [inline field/TinyLang/Field/Evaluator.hs:335]: It was used previously in the compiler, but I think it's fine to delete it now.

File: field/TinyLang/Field/Raw/Parser.hs [hunk 2]
```diff
@@ -141,9 +142,10 @@ statement ::=
     "let" var "=" expr
     "assert" expr
     "for" var "=" int-literal "to" int-literal "do" statements "end"
+    "ext" var
 
 statements ::=
-    (statement (";" statement)*)?
+    (statement ";")*
```
Comment (effectfully) [inline field/TinyLang/Field/Raw/Parser.hs:144]: We allow empty programs? It's fine I guess.

File: field/TinyLang/Field/Typed/TypeChecker.hs [hunk 3]
```diff
@@ -131,20 +125,30 @@ mkSomeUniVar var
...
-    :: forall m f. (MonadTypeChecker m, TextField f)
-    => R.Expr R.Var f -> m (T.SomeUniExpr f)
+inferExpr ::
+    forall m f. (Monad m, TextField f)
+    => R.Expr R.Var f -> TypeCheckerT m (T.SomeUniExpr f)
 inferExpr (R.EConst (Some c@(T.UniConst uni _))) =
     pure $ SomeOf uni $ T.EConst c
-inferExpr (R.EVar   v) = do
+inferExpr (R.EVar v) = do
+    ensureBound v
```
Comment (effectfully) [inline field/TinyLang/Field/Typed/TypeChecker.hs]: We will ensure that `v` is bound by looking up its type once in the typing environment in `inferUniVar`.

File: field/TinyLang/Field/Typed/TypeChecker.hs [hunk 4]
```diff
@@ -65,53 +59,53 @@ type MonadTypeError   m = MonadError TypeCheckError m
...
                           , MonadScope m
                           , MonadTypeError m
+                          , MonadReader TyEnv m
                           )
 
+{-| == Type Environments
+-}
+
+-- TODO: Can be parameterised later
+type TyEnv = Set R.Var
```
Comment (effectfully) [inline field/TinyLang/Field/Typed/TypeChecker.hs]: Oh, I was thinking of a different thing when I heard "typing environment". I thought, we were going to map variables to their types, not just store the variables that are currently in scope. Or is it planned for another PR?

File: field/TinyLang/Field/Typed/TypeChecker.hs [hunk 5]
```diff
@@ -65,53 +59,53 @@ type MonadTypeError   m = MonadError TypeCheckError m
...
+
+
 {-| @TypeChecker@ Transformer
 -}
-newtype TypeCheckerT e (m :: Type -> Type) a =
-    TypeChecker { runTypeCheckerT :: (ExceptT e (StateT Scope (SupplyT m))) a }
+newtype TypeCheckerT (m :: Type -> Type) a =
+    TypeChecker { runTypeCheckerT :: (ExceptT TypeCheckError (StateT Scope (ReaderT TyEnv (SupplyT m)))) a }
```
Comment (effectfully) [inline field/TinyLang/Field/Typed/TypeChecker.hs]: Are we modelling scopes here with both `StateT` and `ReaderT`?

File: field/TinyLang/Field/Typed/TypeChecker.hs [hunk 6]
```diff
@@ -186,71 +186,65 @@ withTypedBinOp R.BAt k = k T.BAt
...
+    withVar var $ kont [T.ELet uniVar tM]
+checkStatement (R.EAssert m) kont = do
+    tM <- checkExpr m
+    kont [T.EAssert tM]
+checkStatement (R.EFor var start end stmts) kont = do
+    T.Statements tStmts <- withVar var $ checkStatements stmts pure
+    tVar <- makeVar $ R.unVar var
+    let uVar = T.UniVar Field tVar
+    let iter i ikont = ikont $ T.ELet uVar (T.EConst . fromIntegral $ i) : tStmts
+    runContT (foldMapA (ContT . iter) [start .. end]) kont
```
Comment (effectfully) [inline field/TinyLang/Field/Typed/TypeChecker.hs]: So as we discussed previously, we can't handle loops like

```
for i = 1 to 3 do
    for j = 1 to i do
       <...>
    end;
end;
```

I.e. we need to map a loop variable to its value either during type checking or before (e.g. we could have a separate pass for unrolling loops in the untyped language). Is it planned for a future PR?

File: hie.yaml [hunk 7]
```diff
@@ -0,0 +1,12 @@
+cradle:
+  stack:
```
Comment (effectfully) [inline hie.yaml:2]: Moved from `dante` to HIE? How's your experience?

**Review by effectfully — COMMENTED** [review]
> This should fix some problems with #76

I thought that problem is already fixed? The problem was that we were renaming programs with loops and we were implicitly assuming that a loop always evaluates at least once, so every variable bound in the loop body gets introduced into the scope. But that was a false assumption, so we removed loops from the typed language. No loops in the typed language = no scoping problems.

There's still ambiguity in the untyped language, but that doesn't matter, at the moment at least.

> but it doesnt have assignments yet.

What do you mean by assignments?

File: field/TinyLang/Field/Evaluator.hs (hunk 1 above)
Comment (effectfully) [reply field/TinyLang/Field/Evaluator.hs:335]: And it was used in the parser, because that is how we were unrolling loops. By inlining the value of a loop variable.

File: field/TinyLang/Field/Typed/TypeChecker.hs (hunk 3 above)
Comment (jakzale) [reply field/TinyLang/Field/Typed/TypeChecker.hs]: Yes, that is the next step, after adding proper type declarations.

File: field/TinyLang/Field/Typed/TypeChecker.hs (hunk 5 above)
Comment (jakzale) [reply field/TinyLang/Field/Typed/TypeChecker.hs]: I am using `ReaderT TyEnv` for modeling scopes using I am using, I think we are using `StateT Scope` for modelling the scope that is later exported to the caller.

File: field/TinyLang/Field/Evaluator.hs (hunk 1 above)
Comment (jakzale) [reply field/TinyLang/Field/Evaluator.hs:335]: Ok.

**Comment by jakzale** [comment]
> I thought that problem is already fixed? The problem was that we were renaming programs with loops and we were implicitly assuming that a loop always evaluates at least once, so every variable bound in the loop body gets introduced into the scope. But that was a false assumption, so we removed loops from the typed language. No loops in the typed language = no scoping problems.

Now we have proper scoping rules for for loops, meaning that the discussion for the value of the for loop variable is moot, as the following program is ill typed
```
for i = 1 to 10 do end;
assert i == 10;
```

> There's still ambiguity in the untyped language, but that doesn't matter, at the moment at least.

What sort of ambiguity you refer to?

> What do you mean by assignments?

Since we add a scope for the for loop, the following programs are ill typed
```
for i = 1 to 10 do
  assert j > 0;
  let j = i * 2;
end;
```
and
```
for i = 1 to 10 do
  let j = i * 2;
end;
assert j == 20;
```

Furthermore, the last assertion in the following program will be true
```
let j = 0;
for i = 1 to 1 do
  let j = i;
end;
assert j == 0;
```

I want to add an assignment operator to modify a variable without adding a new binding, which would make the following program's assertion true
```
let j = 0;
for i = 1 to 1 do
  j <- i;
end;
assert j == 1;
```

File: hie.yaml (hunk 7 above)
Comment (jakzale) [reply hie.yaml:2]: properly configured HIE is faster than dante.  HIE's main problem seems to be the file layout of our codebase.  Mutlipackage configuration seems to fix it (partially).

HIE provides much well-rounded experience, with built-in HLint giving tips on how to improve existing codebase.  It provides slightly more information when inspecting symbols, which is nice.

I think that currently on Emacs `dante` is better, because:
- HIE can lag behind some edits, compaining about an error that is already fixed,
- HIE gets confused with our codebase (it doesn't seem to properly detect contents of the cabal file).  Probably the root directory is not picked properly.  I wasn't able to fix that yet.  HIE with NeoVim/VSCode works fine, so I presume that there is some bug somewhere in one of emacs lsp packages that I am using.

Overall, I think that HIE is a step in the right direction, but I do not think it provides much added value over `dante` (yet).

What I actually would want, is a way to open a local haddock page for a symbol I am using, or a way to perform a quick local hoogle search.  I presume HIE supports at least the latter, but I wasn't bothered enough to implement it myself (for our codebase `rg` seems fine for me).

File: field/TinyLang/Field/Typed/TypeChecker.hs (hunk 6 above)
Comment (jakzale) [reply field/TinyLang/Field/Typed/TypeChecker.hs]: Yes, I plan to add it as a separate PR.

I will probably add it as a type checking mechanism.  Ideally I want to avoid interleaving of normalisation and typechecking or evaluation and typechecking.

I also wanted to read a bit about constant folidng and constant propagation first.

File: field/TinyLang/Field/Typed/TypeChecker.hs (hunk 4 above)
Comment (jakzale) [reply field/TinyLang/Field/Typed/TypeChecker.hs]: Yes, next PR will switch from intrinsically typed variables to extrinsically typed variables.

File: field/TinyLang/Field/Typed/TypeChecker.hs (hunk 5 above)
Comment (effectfully) [reply field/TinyLang/Field/Typed/TypeChecker.hs]: Ah, indeed. I think the caller only needs mappings from external variable names to variable IDs. Let's keep it the way it's currently written, but I'll need to rethink it later.

Comment (effectfully) [reply field/TinyLang/Field/Typed/TypeChecker.hs]: Ok, it makes sense. We only use `Scope` to ensure that a variable name always maps to the same variable, completely regardless of scoping. Scopes are resolved later in the renamer.

But then we now also resolve scopes in the type checker in addition to resolving them in the renamer. And we can't apply the renamer beforehand, because of loops.

Would the best way be to

1. unroll loops
2. rename (after moving the renamer to operate on untyped terms rather than typed ones)
3. type check

? Ok, I'll think about that.

File: field/TinyLang/Field/Typed/TypeChecker.hs (hunk 6 above)
Comment (effectfully) [reply field/TinyLang/Field/Typed/TypeChecker.hs]: > I will probably add it as a type checking mechanism. Ideally I want to avoid interleaving of normalisation and typechecking or evaluation and typechecking.

Yeah, it would be best to avoid interleaving (partial) evaluation and type checking.

File: hie.yaml (hunk 7 above)
Comment (effectfully) [reply hie.yaml:2]: Thanks for the info.

> What I actually would want, is a way to open a local haddock page for a symbol I am using,

Jump-to-definition, then jump back?

Comment (jakzale) [reply hie.yaml:2]: Does jump to definition jump to source code of a dependency?

Comment (effectfully) [reply hie.yaml:2]: Yeah, if [configured right](https://github.com/effectfully/sketches/tree/master/haskell-emacs-jump-to-definition).

**Comment by effectfully** [comment]
> What sort of ambiguity you refer to?

I misunderstood what you're after.

I've responded in the [Loops](https://github.com/effectfully/tiny-lang/issues/54) issue.

File: field/TinyLang/Field/Typed/TypeChecker.hs [hunk 8]
```diff
@@ -186,71 +186,65 @@ withTypedBinOp R.BAt k = k T.BAt
...
+    tM <- T.withKnownUni uni $ checkExpr m
+    withVar var $ kont [T.ELet uniVar tM]
+checkStatement (R.EAssert m) kont = do
+    tM <- checkExpr m
+    kont [T.EAssert tM]
+checkStatement (R.EFor var start end stmts) kont = do
+    T.Statements tStmts <- withVar var $ checkStatements stmts pure
+    tVar <- makeVar $ R.unVar var
+    let uVar = T.UniVar Field tVar
+    let iter i ikont = ikont $ T.ELet uVar (T.EConst . fromIntegral $ i) : tStmts
```
Comment (effectfully) [inline field/TinyLang/Field/Typed/TypeChecker.hs]: So in the typed language `uVar` is in scope after the loop?

Comment (jakzale) [reply field/TinyLang/Field/Typed/TypeChecker.hs]: Yes and No.

Yes, because we unroll a loop to a series of let bindings, meaning that technically every variable introduced is in scope.

No, because it inaccessible after the loop.  The following program will not type check
```
for x = 0 to 0 do; end;
assert x == 0;
```

Comment (effectfully) [reply field/TinyLang/Field/Typed/TypeChecker.hs]: > No, because it inaccessible after the loop. The following program will not type check
> ```
> for x = 0 to 0 do; end;
> assert x == 0;
> ```

Hmm, okay. What does

```
let x = 1;
for x = 0 to 0 do; end;
assert x == 0;
```

elaborate to? Looking at the current version with `withVar` I feel like this would elaborate to the wrong thing, but not sure. Could you add such a golden test?

File: common/TinyLang/Environment.hs [hunk 9]
```diff
@@ -58,3 +59,6 @@ fromVars = toEnvBy $ uncurry insertVar
 
 toUniques :: Env a -> [(Unique, a)]
 toUniques = map (first Unique) . IntMap.toList . unEnv
+
+toVars :: Env a -> [a]
+toVars =  IntMap.elems . unEnv
```
Comment (effectfully) [inline common/TinyLang/Environment.hs]: Why calling it `toVars` if it doesn't return any `Var`s? Maybe just `elems` like the corresponding `IntMap.elems`?

File: field/TinyLang/Field/Core.hs [hunk 10]
```diff
@@ -14,6 +15,14 @@ newtype Statements stmt = Statements { unStatements :: [stmt] }
...
 
 -- | Basic wrapper of program
-newtype Program stmt = Program { unProgram :: Statements stmt }
-    deriving (Generic, Eq, Functor, Foldable, Traversable)
-    deriving (Show) via (Quiet (Program stmt))
+data Program var stmt = Program { unExts :: [var], unProgram :: Statements stmt }
+    deriving (Generic, Eq, Foldable, Traversable)
+
+instance Functor (Program vars) where
+    fmap f (Program vars stmts) = Program vars (fmap f stmts)
```
Comment (effectfully) [inline field/TinyLang/Field/Core.hs]: I'd just derive it.

File: field/TinyLang/Field/Core.hs [hunk 11]
```diff
@@ -14,6 +15,14 @@ newtype Statements stmt = Statements { unStatements :: [stmt] }
     deriving (Show) via (Quiet (Statements stmt))
 
 -- | Basic wrapper of program
-newtype Program stmt = Program { unProgram :: Statements stmt }
-    deriving (Generic, Eq, Functor, Foldable, Traversable)
-    deriving (Show) via (Quiet (Program stmt))
+data Program var stmt = Program { unExts :: [var], unProgram :: Statements stmt }
```
Comment (effectfully) [inline field/TinyLang/Field/Core.hs]: We employ the convention that `unX` is only used for `newtype`s. Now that `Program` is a `data`, it should be something like

```haskell
data Program var stmt = Program
    { _programExts :: [var]
    , _programStatements :: Statements stmt
    }
```

Should we maybe call the `var` parameter `ext`? 'Cause this is more precise and more general at the same time. And we probably want to instantiate that parameter with a variable paired with its type, which is a bit more than just a variable.

File: field/TinyLang/Field/Generator.hs [hunk 12]
```diff
@@ -106,23 +103,34 @@ adjustUniquesForVars :: MonadSupply m => Vars f -> m ()
...
-    vectorVars <- traverse (make Vector) ["#q", "#r", "#s", "#t", "#u", "#v", "#w"]
-    return $ fieldVars ++ boolVars ++ vectorVars
 
 -- | A wrapper around @UniVar f a@ provided for its @Arbitrary@ instance that allows to generate
 -- variables from the default set of them.
-newtype DefaultUniVar f a = DefaultUniVar
-    { unDefaultUniVar :: UniVar f a
+
+defaultIdents :: [String]
+defaultIdents = concat . transpose $ [fieldIdents, boolIdents, vectorIdents] where
```
Comment (effectfully) [inline field/TinyLang/Field/Generator.hs:111]: Nice.

File: field/TinyLang/Field/Generator.hs [hunk 13]
```diff
@@ -106,23 +103,34 @@ adjustUniquesForVars :: MonadSupply m => Vars f -> m ()
...
+type DefaultVars          = Default [Var]
+type DefaultSomeUniVars f = Default (Vars f)
+
+instance Arbitrary DefaultVars where
+    arbitrary = do
+        vars <- sublistOf $ defaultIdents
+        pure . Default $ runSupply $ traverse freshVar vars
+
+instance Arbitrary (DefaultSomeUniVars f) where
+    arbitrary = Default . fmap mkSomeUniVar . unDefault <$> arbitrary
```
Comment (effectfully) [inline field/TinyLang/Field/Generator.hs]: Would

```haskell
arbitrary = traverse mkSomeUniVar <$> arbitrary
```

work?

File: field/TinyLang/Field/Generator.hs [hunk 14]
```diff
@@ -288,7 +299,7 @@ boundedArbitraryStmt size
                                let vars' = Some uniVar : vars
                                    size' = size - 1
                                put vars'
-                               ELet uniVar <$> boundedArbitraryExpr vars' size')
+                               ELet uniVar <$> boundedArbitraryExpr vars size')
```
Comment (effectfully) [inline field/TinyLang/Field/Generator.hs:302]: Why the change?

File: field/TinyLang/Field/Generator.hs [hunk 15]
```diff
@@ -416,16 +428,34 @@ defaultUniConst =
...
 instance (Field f, Arbitrary f) => Arbitrary (Program f) where
-    arbitrary = Program <$> arbitrary
-    shrink    = fmap Program . shrink . unProgram
+    arbitrary = do
+        vars  <- unDefault <$> arbitrary
+        let someUniVars = fmap mkSomeUniVar vars
+        stmts <- runSGen someUniVars $ sized $ \size -> do
+                     adjustUniquesForVars someUniVars
+                     boundedArbitraryStmts size
+        -- NOTE:  abritrary statements uses
```
Comment (effectfully) [inline field/TinyLang/Field/Generator.hs]: Don't get the NOTE.

File: field/TinyLang/Field/Generator.hs [hunk 16]
```diff
@@ -416,16 +428,34 @@ defaultUniConst =
...
+        pure $ Program vars stmts
+
+    shrink (Program vars stmts) =
+        fmap (uncurry Program) $
+            -- [ (vars', stmts)  | vars'  <- shrinkList (const []) vars ] ++
+            [ (vars , stmts') | stmts' <- shrink stmts ]
```
Comment (effectfully) [inline field/TinyLang/Field/Generator.hs]: Commented out code. Should we prune `vars` if they are not referenced by `stmt'`?

File: field/TinyLang/Field/Generator.hs [hunk 17]
```diff
@@ -505,14 +535,22 @@ genEnvFromVarSigs =
     traverse $ \(VarSig _ (uni :: Uni f a)) ->
         Some <$> withKnownUni uni (arbitrary :: Gen (UniConst f a))
 
+genInputEnvFromExts :: (Field f, Arbitrary f) => [Var] -> Gen (Env (SomeUniConst f))
+genInputEnvFromExts vars = fromVars . zip vars <$> consts where
+    consts = traverse constFromVar vars
+    constFromVar tVar =
```
Comment (effectfully) [inline field/TinyLang/Field/Generator.hs]: I'd write this as

```haskell
consts =
    for vars $ \tvar -> 
        case uniOfVar . _varName $ tVar of
            Some (uni :: Uni f a) ->
                Some <$> withKnownUni uni (arbitrary :: Gen (UniConst f a))
```

but not too important.

File: field/TinyLang/Field/Rename.hs [hunk 18]
```diff
@@ -9,9 +9,12 @@ import           TinyLang.Field.Typed.Core
 import           Control.Monad.Cont
 
 renameProgram :: MonadSupply m => Program f -> m (Program f)
-renameProgram (Program stmts) = do
-    stmtsSupplyFromAtLeastFree stmts
-    Program <$> runRenameM (withRenamedStatementsM stmts pure)
+renameProgram prog@(Program vars stmts) = do
+    progSupplyFromAtLeastFree prog
```
Comment (effectfully) [inline field/TinyLang/Field/Rename.hs:15]: Please add a note that we need to revise this in future, because we probably don't need to handle programs with free variables in the compiler and the renamer runs right before the main compilation pass starts.

File: field/TinyLang/Field/Typed/TypeChecker.hs [hunk 19]
```diff
@@ -44,107 +33,94 @@ import qualified TinyLang.Field.Typed.Core  as T
...
-    | '?':_ <- _varName var = Some $ T.UniVar Bool   var
-    | '#':_ <- _varName var = Some $ T.UniVar Vector var
-    | otherwise             = Some $ T.UniVar Field  var
+-- NOTE:  At the moment this mimics the old scope
+-- TODO:  Discuss wether we should add an env on top of @withVar@ or not.
+withVar :: (Monad m) => R.Var -> forall r. (T.Var -> TypeCheckerT m r) -> TypeCheckerT m r
+withVar var kont = do
+    tyEnv <- ask
+    case Map.lookup var tyEnv of
+        Just tVar -> kont tVar
```
Comment (effectfully) [inline field/TinyLang/Field/Typed/TypeChecker.hs:102]: This doesn't look right. The point of `MonadScope` was to assign the same unique to all variables that share a name. Regardless of whether those variables are distinct or not. That logic doesn't work for mapping untyped variables to typed ones as here we need to track scopes properly.

In particular,

```
let x = 0;
let x = 1;
assert x == 1;
```

now elaborates to the wrong

```
let x_0 = 0;
let x_1 = 1;
assert x_0 == 1;
```

which is obscured by the fact that we recover the correct scoping in the renamer.

This suggests that we should have two kinds of golden tests: the program that we get right after type checking and the program that we get after renaming. Unless I'm wrong, it'll show that we now do weird things in the type checker.

File: test/Field/golden/01-uniques.golden [hunk 20]
```diff
@@ -1,10 +1,11 @@
-let a_1 = a_0;
+ext a_1;
```
Comment (effectfully) [inline test/Field/golden/01-uniques.golden]: Why do we start enumerating variables with `1` rather than `0` now?

File: test/Field/golden/03-issue-76.golden [hunk 21]
```diff
@@ -0,0 +1,2 @@
+ext x_2;
+let y_3 = if (0 == x_2) then 0 else 0;
```
Comment (effectfully) [inline test/Field/golden/03-issue-76.golden:2]: Ok, so we can conclude that #76 is fixed.

File: field/TinyLang/Field/Typed/TypeChecker.hs (hunk 19 above)
Comment (effectfully) [reply field/TinyLang/Field/Typed/TypeChecker.hs:102]: > which is obscured by the fact that we recover the correct scoping in the renamer.

Hm, I looked into the renamer and I don't think it can recover scoping. I'm really confused now, how is this whole thing working? Golden tests suggest that we do have correct scoping, but looking at the code I don't see how that can possibly be the case. Can we have a unit test with 

```
let x = 0;
let x = 1;
assert x == 1;
```

where we only type check things and don't rename them?

Comment (effectfully) [reply field/TinyLang/Field/Typed/TypeChecker.hs:102]: Oh, I think I see what's going on. So taking

```
let x = 0;
let x = 1;
assert x == 1;
```

we call

```haskell
checkStatement (R.ELet var m) kont =
    case uniOfVar . R.unVar $ var of
        Some uni -> do
                tM <- T.withKnownUni uni $ checkExpr m
                withVar var $ \ tVar -> kont [T.ELet (UniVar uni tVar) tM]
```

over the first and then the second let, so we extend the initially empty environment first, but then we don't extend it and instead use the already existing binding. So the program elaborates to

```
let x_0 = 0;
let x_0 = 1;
assert x_0 == 1;
```

which then gets renamed correctly.

I'm not entirely sure if it should be causing us any problems right now, but it certainly will in future. For example, we'll need to inline the value of a loop variable and if the loop variable has the same name than some previously referenced variable, then we can't just use the value of that variable, we have to actually update the environment instead of reusing the existing entry that just became shadowed. Or in future we may have stuff like

```
let x = T;
let x = 1;
assert x == 1;
```

and here we also need to update the environment and can't reuse the existing entry, because the type of the new binding differs from the type of the old one.

File: field/TinyLang/Field/Core.hs (hunk 11 above)
Comment (jakzale) [reply field/TinyLang/Field/Core.hs]: > And we probably want to instantiate that parameter with a variable paired with its type, which is a bit more than just a variable.

I am leaving that for the next PR, that will switch from intrinsincally typed variables to extrinsically typed variables.

File: field/TinyLang/Field/Generator.hs (hunk 13 above)
Comment (jakzale) [reply field/TinyLang/Field/Generator.hs]: No, because `traverse` will collect the side effect of `[]` and we will end up with `[Default a]` not `Default [a]`.

If we want to play code golf, it is:
``` haskell
arbitrary = (fmap . fmap . fmap) mkSomeUniVar arbitrary
```

File: test/Field/golden/03-issue-76.golden (hunk 21 above)
Comment (jakzale) [reply test/Field/golden/03-issue-76.golden:2]: It looks like it.  I would keep it as a regression test though.

File: test/Field/golden/01-uniques.golden (hunk 20 above)
Comment (jakzale) [reply test/Field/golden/01-uniques.golden]: Ah, this looks like a bug.

Comment (jakzale) [reply test/Field/golden/01-uniques.golden]: Hmm, `renameProgram` seems to consume `MonadSupply` for each variable.  For instance
```
ext a;
ext b;
...
```
will start numbering variables from `2`.  I need to check if that is indeed a bug.

Comment (jakzale) [reply test/Field/golden/01-uniques.golden]: I am not sure if it is a bug.  The above behaviour is due to the fact that we share `MonadSupply` between `typeProgram` and `renameProgram` in `parseProgramFrom`.  Since we expose `MonadSupply` in `parseProgramFrom`, I presume that it is not a bug, just us being extremely careful.

Here is the explanation:

1. When typechecking program for the first time, we assign a fresh unique for each distinct identifier, meaning that the following program:

```
ext a;
ext b;
let a = a;
```

get typechecked as

```
ext a_0;
ext b_1;
let a_0 = a_0;
```

then we do the renaming step, which matches uniques to scopes:

```
ext a_2;
ext b_3;
let a_4 = a_2;
```

File: field/TinyLang/Field/Typed/TypeChecker.hs (hunk 19 above)
Comment (jakzale) [reply field/TinyLang/Field/Typed/TypeChecker.hs:102]: Ok, this actually gave me slight headache as well, because there are two notions of environment conflated into one:
- tracking which uniques are assigned to which identifier, and
- tracking which variables are in scope.

I do not think that untangling them together should be an issue.

# Explanation

1. I refactored `withVar` to pass the typed version of the variable to the continuation in order to encapsulate the process of changing from `R.Var` to `T.Var`.
2. To keep the old behaviour of giving the same unique to the same identifier, we use a `Map R.Var T.Var` to track _which uniques are assigned to which identifiers_.
3. We reuse the same map to track which variables are in scope.
4. As variables are intrinsically typed, we use the name of variable to infer its type/uni.

Once we switch to extrinsically-typed variables, we will probably need to:
- Use a different map for tracking which uniques have been assigned (`Map (R.Var, Uni) T.Var`)
- Use a proper typing environment, which similar to `Map R.Var (T.Var, Uni)`.

File: field/TinyLang/Field/Core.hs (hunk 11 above)
Comment (effectfully) [reply field/TinyLang/Field/Core.hs]: > And we probably want to instantiate that parameter with a variable paired with its type, which is a bit more than just a variable.

Yeah, that's only an argument in favor of renaming `var` to `ext`. Not suggesting to pair variables with types in this PR.

File: field/TinyLang/Field/Generator.hs (hunk 13 above)
Comment (effectfully) [reply field/TinyLang/Field/Generator.hs]: Ah, yeah, `mkSomeUniVar` is not monadic. But then 

```haskell
arbitrary = fmap mkSomeUniVar <$> arbitrary
```

should work?

```haskell
arbitrary = (fmap . fmap . fmap) mkSomeUniVar arbitrary
```

Do you need that many `fmap`s?

File: test/Field/golden/01-uniques.golden (hunk 20 above)
Comment (effectfully) [reply test/Field/golden/01-uniques.golden]: Well, not a big deal, I guess. I do need uniques to start from `0` for golden tests, but I do a separate renaming pass in the compiler anyway, so `MonadSupply` is not shared with anything there.

File: test/Field/golden/03-issue-76.golden (hunk 21 above)
Comment (effectfully) [reply test/Field/golden/03-issue-76.golden:2]: Sure.

File: field/TinyLang/Field/Typed/TypeChecker.hs (hunk 19 above)
Comment (effectfully) [reply field/TinyLang/Field/Typed/TypeChecker.hs:102]: Previously we had this:

> tracking which uniques are assigned to which identifier, and

and not this:

> tracking which variables are in scope.

Now that we have the latter and given that I can always get the list of external variables of a program, I don't think we need the former anymore. So once we switch to extrinsically-typed variables I propose to only have `Map R.Var (T.Var, Uni)` (or `Map R.Var T.UniVar` for that matter).

Or do you have a use case for `Map (R.Var, Uni) T.Var`?

Comment (jakzale) [reply field/TinyLang/Field/Typed/TypeChecker.hs:102]: > Or do you have a use case for `Map (R.Var, Uni) T.Var`?
Not really.

Given that we relax our requirement for uniques, maybe we should parameterise the typed ast by the type of variable as well.  Personally I would like to work without uniques and supply.  Especially, since we will perform a renaming step in the compiler anyways.

File: field/TinyLang/Field/Typed/TypeChecker.hs (hunk 8 above)
Comment (jakzale) [reply field/TinyLang/Field/Typed/TypeChecker.hs]: I am resolving this conversation as we went back to the previous notion of loops, where the loop variable is bound after the loop.

Comment (effectfully) [reply field/TinyLang/Field/Typed/TypeChecker.hs]: > I am resolving this conversation as we went back to the previous notion of loops, where the loop variable is bound after the loop.

Is this something that we're going to have in the long run? If the plan is to inline the value of a loop variable, then we can unscope the loop variable after the loop (and only inline in the body of the loop then).

**Comment by jakzale** [comment]
>
> Is this something that we're going to have in the long run? If the plan is
> to inline the value of a loop variable, then we can unscope the loop
> variable after the loop (and only inline in the body of the loop then).
>
Hmm, I need to think about it.  Yes, we could technically inline the loop
variable and erase it completely.
I am still not sure how to do it properly, as there are programs in the
spec that depend on variables bound in the body of a for loop.
Lemme think about it for a bit.

**Comment by effectfully** [comment]
> I am still not sure how to do it properly, as there are programs in the
> spec that depend on variables bound in the body of a for loop.

Why is that a problem?

```
for i = 0 to 1 do
  let x = i;
assert x == 1;
```

will elaborate to

```
let x = 0;
let x = 1;
assert x == 1;
```

**Comment by jakzale** [comment]
Hmm, maybe I am overdoing it.

File: field/TinyLang/Field/Generator.hs (hunk 14 above)
Comment (jakzale) [reply field/TinyLang/Field/Generator.hs:302]: Ah, because `vars'` contains `uniVar`, which is not bound in the body of the let binding.

File: field/TinyLang/Field/Generator.hs (hunk 15 above)
Comment (jakzale) [reply field/TinyLang/Field/Generator.hs]: Expanded the NOTE.  Using `Arbitrary` instance of `Statements f` can potentially generate open programs.

File: field/TinyLang/Field/Generator.hs [hunk 22]
```diff
@@ -416,16 +428,35 @@ defaultUniConst =
...
+                     adjustUniquesForVars someUniVars
+                     boundedArbitraryStmts size
+        -- NOTE: we should not use arbitrary instance for Statements f directly,
+        -- as it can generate a statement that contains free variables.
+        pure $ Program exts stmts
+
+    shrink (Program exts stmts) =
+        fmap (uncurry Program) $
+            -- [ (vars', stmts)  | vars'  <- shrinkList (const []) vars ] ++
+            [ (exts , stmts') | stmts' <- shrink stmts ]
```
Comment (effectfully) [inline field/TinyLang/Field/Generator.hs:451]: It would make sense to remove those external variables that are no longer referenced in `stmts'`.

File: field/TinyLang/Field/Typed/TypeChecker.hs (hunk 19 above)
Comment (effectfully) [reply field/TinyLang/Field/Typed/TypeChecker.hs:102]: This still does not look right.

Why do we need the 

```haskell
    case Map.lookup var tyEnv of
        Just tVar -> kont tVar
```

part? We needed this kind of logic previously, but I'm currently failing to see why we need it now.

What happens if we remove that and do

```haskell
            tVar <- freshVar . R.unVar $ var
            local (Map.insert var tVar) $ kont tVar
```

(the `Nothing` case) unconditionally?

**Comment by jakzale** [comment]
Ah yes, we can do that, then we do not need to do renaming in type checking.

File: field/TinyLang/Field/Generator.hs (hunk 22 above)
Comment (jakzale) [reply field/TinyLang/Field/Generator.hs:451]: Fixed in 19918bf535ab6cd454af67eccfcf48768202e0da, which is part of #79
